ghga_service_chassis_lib/s3.py

```diff
@@ -172,6 +172,46 @@ def _translate_s3_client_errors(
     return exception
 
 
+def adapt_part_size(current_part_size: int, file_size: int) -> int:
+    """
+    Checks if the provided part size is within bounds and produces <= 10_000 parts.
+
+
+    Args:
+        current_part_size (int): currently chosen part size in bytes to be checked
+        file_size (int): total file size in bytes
+
+    Raises:
+        ValueError if file size exceeds the maximum of 5 TiB
+    """
+    lower_bound = 5 * 1024**2
+    upper_bound = 5 * 1024**3
+    hard_upper_limit = 5 * 1024**4
+    max_num_parts = 10_000
+
+    if file_size > hard_upper_limit:
+        raise ValueError(
+            f"""Provided file size of {file_size/1024**4:2.f}TiB
+            exceeds maximum allowed file size of 5 TiB"""
+        )
+
+    # clamp to lower/upper bound, respectively
+    if current_part_size < lower_bound:
+        current_part_size = lower_bound
+    elif current_part_size > upper_bound:
+        current_part_size = upper_bound
+
+    # powers of two from 8 MiB to 1024 MiBs, everything above would produce files
+    # over the limit
+    part_size_candidates = [2**i * 1024**2 for i in range(3, 11)]
+
+    if file_size / current_part_size > max_num_parts:
+        for part_size_candidate in part_size_candidates:
+            if file_size / part_size_candidate <= 10_000:
+                return part_size_candidate
+    return current_part_size
```

Comment:
I renamed the current_part_size above to preferred_part_size, in case you except that suggestions, the following need to change as well:
```suggestion
    if preferred_part_size < lower_bound:
        preferred_part_size = lower_bound
    elif preferred_part_size > upper_bound:
        preferred_part_size = upper_bound

    # powers of two from 8 MiB to 1024 MiBs, everything above would produce files
    # over the limit
    part_size_candidates = [2**i * 1024**2 for i in range(3, 11)]

    if file_size / preferred_part_size > max_num_parts:
        for part_size_candidate in part_size_candidates:
            if file_size / part_size_candidate <= 10_000:
                return part_size_candidate
    return preferred_part_size
```